### DISCUSSION TITLE
State Management

### DISCUSSION THREAD

**Discussion by seocwen** [nathanhoad/godot_dialogue_manager#196] [General]
I wish it were a little easier to manage state using this system. For instance, having "const" and "var" keywords. The most obvious use of const would be to create simple string variables for including snatches of text smaller than a line. If those const strings could also use the dialogue managers syntax and include their own variable references, it could be very powerful. Eg:

```
const cute_animals = "[[kitties|puppies|bunnies]]"
const judas_epithet = "bright eyed Judas, the pet store owner, petter of {{happy_animals}}"

~ pet_store_story
    I saw before me {{judas_epithet}}, with a crazed look in his eye. He was holding some {{cute_animals}} over a pot of boiling stew. I screamed at him, "You are {{judas epithet}}, snap out of it or the the demon that's possessed you will have your soul!" But it was to no avail. I consoled myself by weeping into some {{cute_animals}}.
```

Other types of variables like integers and so on might also come in handy to prevent magic numbers. A similar var keyword to create a temporary variable at the current indent level would be similarly useful, rather than having to juggle state in and out of an external global. Finally, and I know this might seem like a much more bloated thing to ask for, but it would also be useful if you could pass variables to nodes when jumping. Eg.

```
~ introduction(name, color)
    {{name}}: Hi, my name is {{name}} and my favorite color is {{color}}

~ ben_arrives
    =>< introduction("Ben", "blue)
    Ben: I am skillful in ice-cream wizardry, but weak against the cold. It's a tough life.
```

**Comment by nathanhoad** [comment]
The stateless nature of the dialogue manager is a deliberate design decision. If you're looking for a more stateful dialogue system then you should give [Dialogic](https://github.com/coppolaemilio/dialogic) or [Ink](https://github.com/paulloz/godot-ink) a whirl.

**Reply by seocwen** [reply]
Fair enough. I guess so it's a lighter system?

Something like constants should be stateless at least. Mostly I'd like to be able to insert text within the same line, instead of advancing to the next textbox. Probably existing titles would work if there was syntax to allow lines to, instead of advancing the textbox, to break the last line, or to append to the last line.

Looking into it, Ink seems to use "<>" for this purpose. I'd consider just using Ink, but the 4.0 version doesn't seem to be working yet, so I'll probably just hook into global methods to handle state for now. Not too hard, really.
